File: LarkCardKit/TemplateHelper.cs
using LarkCardKit.Models;
using LarkCardKit.Models.Elements;
using LarkCardKit.Templates;

namespace LarkCardKit;

/// <summary>
/// 模板参数填充工具类
/// 支持在卡片文本内容中使用 ${变量名} 格式的模板变量
/// </summary>
public static class TemplateHelper
{
    /// <summary>
    /// 使用参数字典填充卡片中的模板参数
    /// </summary>
    /// <param name="card">卡片对象</param>
    /// <param name="parameters">参数字典</param>
    public static void FillTemplate(this Card card, Dictionary<string, object?> parameters)
    {
        var filler = new TemplateParameterFiller();
        filler.SetParameters(parameters);
        card.FillTemplate(filler);
    }

    /// <summary>
    /// 使用 TemplateParameterFiller 填充卡片中的模板参数
    /// </summary>
    /// <param name="card">卡片对象</param>
    /// <param name="filler">模板参数填充器</param>
    public static void FillTemplate(this Card card, TemplateParameterFiller filler)
    {
        if (card.Header != null)
        {
            FillTemplateInElement(card.Header.Title, filler);
            FillTemplateInElement(card.Header.Subtitle, filler);
        }

        FillTemplateInElements(card.Body.Elements, filler);
    }

    /// <summary>
    /// 在元素中填充模板参数
    /// </summary>
    private static void FillTemplateInElement(Element? element, TemplateParameterFiller filler)
    {
        if (element == null) return;

        switch (element)
        {
            case PlainText plainText:
                plainText.Content = filler.FillString(plainText.Content);
                break;
            case Markdown markdown:
                markdown.Content = filler.FillString(markdown.Content);
                break;
            case MarkdownText markdownText:
                markdownText.Content = filler.FillString(markdownText.Content);
                break;
            case ColumnSet columnSet:
                foreach (var column in columnSet.Columns)
                {
                    FillTemplateInElements(column.Elements, filler);
                }
                break;
            case Form form:
                FillTemplateInElements(form.Elements, filler);
                break;
            case CollapsiblePanel collapsiblePanel:
                if (collapsiblePanel.Header != null)
                {
                    FillTemplateInElement(collapsiblePanel.Header.Title, filler);
                }
                FillTemplateInElements(collapsiblePanel.Elements, filler);
                break;
            case Loop loop:
                FillTemplateInElement(loop.Template, filler);
                break;
        }
    }

    /// <summary>
    /// 在元素列表中填充模板参数
    /// </summary>
    private static void FillTemplateInElements(List<Element> elements, TemplateParameterFiller filler)
    {
        if (elements == null) return;

        foreach (var element in elements)
        {
            FillTemplateInElement(element, filler);
        }
    }
}

File: Python/lark_card_kit/utils/template_helper.py
"""Template helper for card-level template filling."""

from __future__ import annotations

from typing import Any

from ..models.card import Card
from ..models.elements import (
    Element,
    PlainText,
    Markdown,
    MarkdownText,
    ColumnSet,
    Form,
    CollapsiblePanel,
    Loop,
)
from ..templates.template_filler import TemplateParameterFiller


def fill_template(card: Card, filler: TemplateParameterFiller) -> None:
    """Fill template parameters in a card.

    Args:
        card: Card to fill template parameters in.
        filler: Template parameter filler.
    """
    if card.header:
        _fill_template_in_element(card.header.title, filler)
        _fill_template_in_element(card.header.subtitle, filler)

    _fill_template_in_elements(card.body.elements, filler)


def fill_template_with_params(card: Card, parameters: dict[str, Any]) -> None:
    """Fill template parameters in a card using a parameter dictionary.

    Args:
        card: Card to fill template parameters in.
        parameters: Parameter dictionary.
    """
    filler = TemplateParameterFiller()
    filler.set_parameters(parameters)
    fill_template(card, filler)


def _fill_template_in_element(element: Element | None, filler: TemplateParameterFiller) -> None:
    """Fill template parameters in an element.

    Args:
        element: Element to fill template parameters in.
        filler: Template parameter filler.
    """
    if element is None:
        return

    if isinstance(element, PlainText):
        element.content = filler.fill_string(element.content)
    elif isinstance(element, Markdown):
        element.content = filler.fill_string(element.content)
    elif isinstance(element, MarkdownText):
        element.content = filler.fill_string(element.content)
    elif isinstance(element, ColumnSet):
        for column in element.columns:
            _fill_template_in_elements(column.elements, filler)
    elif isinstance(element, Form):
        _fill_template_in_elements(element.elements, filler)
    elif isinstance(element, CollapsiblePanel):
        if element.header:
            _fill_template_in_element(element.header.title, filler)
        _fill_template_in_elements(element.elements, filler)
    elif isinstance(element, Loop):
        _fill_template_in_element(element.template, filler)


def _fill_template_in_elements(elements: list[Element] | None, filler: TemplateParameterFiller) -> None:
    """Fill template parameters in a list of elements.

    Args:
        elements: List of elements to fill template parameters in.
        filler: Template parameter filler.
    """
    if elements is None:
        return

    for element in elements:
        _fill_template_in_element(element, filler)